Basic support for Helm charts

alonsodomin opened the pull request:
This is a follow up from the massive #14487 PR and the discussion on attempting to split it to make it much easier to review.

The changes in here bring the foundational structures used in the previous PR to be able to handle Helm charts _as a single compilation unit_. To prove that this works fine in combination with the Helm commands, the `lint` goal is also implemented.

Test coverage has been improved in comparison to the same code in the previous PR, but the approach with the `helm_chart` target is the same: Instead of defining fine grained targets, the `helm_chart` target is quite coarse as it includes different types of files as its _sources_: the `Chart.yaml` file, the local `values.yaml` file (which brings the defaults), the `templates/*.tpl` files (go partial templates) and the `templates/*.yaml` files (the kuberentes manifest templates using Go template syntax).

A different approach to this, and maybe one that is worth exploring, would be to have the following targets:

`chart/templates/BUILD`:
```python
helm_partial_templates(name="partials")  # Uses default sources="*.tpl"
helm_kube_tempaltes(name="kube") # Uses default sources="*.yaml"
```

`chart/BUILD`:
```python
 # Has two source fields:
#  - chart="Chart.yaml"
# - defaults="values.yaml"
helm_chart(dependencies=["templates:partials", "templates:kube"])
```

What I like a lot with this approach is that each target is very specific and can be configured very easily. What I mostly don't like is the explosion of targets if produces and having to set the `helm_chart` dependencies expicitly - although this could be mitigated with dependency injection/inference ...

Let me know what you think of current approach VS the one mentioned here.

alonsodomin commented on src/python/pants/backend/helm/goals/lint.py:
```diff
@@ -0,0 +1,75 @@
+    charts = await MultiGet(
+        Get(HelmChart, HelmChartFieldSet, HelmChartFieldSet.create(wrapped.target))
+        for wrapped in chart_targets
+    )
```
I shouldn't need to use the `WrappedTarget` rule to perform this, as a `HelmLintFieldSet` is itself a `HelmChartFieldSet`.

However when I attempt to do:

```python
charts = await MultiGet(
  Get(HelmChart, HelmChartFieldSet, field_set)
  for field_set in request.field_sets
)
```

I get a type error as `field_set` is not exactly a `HelmChartFieldSet`. Which makes me think that `Get` is invaliant or basically that I don't know enough Python.

tdyas replied:
Noting to do with Python actually since `Get` is a concept of the Pants engine, not Python generally. The Pants engine is type invariant (except with union types). 

See https://www.pantsbuild.org/docs/rules-api-concepts#valid-types

> Unlike Python in general, the engine uses exact type matches, rather than considering inheritance; even if Truck subclasses Vehicle, the engine will view these types as completely separate when deciding which rules to use.

alonsodomin replied:
forgot to review that part of the docs... then I think that to avoid the usage of the `WrappedTarget`, I will need to change the request type of the rule, something like:

```python
@dataclass(frozen=True)
class HelmChartRequest:
  field_set: HelmChartFieldSet

@rule
async def get_helm_chart(request: HelmChartRequest) -> HelmChart:
 ...
```

And with that I could avoid the `WrappedTarget` with something like:

```python
charts = await MultiGet(
  Get(HelmChart, HelmChartRequest(field_set))
  for field_set in request.field_sets
```

cognifloyd commented:
I recently had to start writing chart unittests with helm-unittest (there are a bunch of forks. This one is presently the most maintained version: https://github.com/quintush/helm-unittest).

It would be nice to use that in the future and it would be an excellent candidate for remote execution because of how it templates things.

`helm-unittest` only renders the minimum required templates for each test. That means that I have to explicitly list the target template file and all templates it depends on (ie templates that are used via `include`). For example, here is a snippet from one of the tests I was working on:

```yaml
templates:
  # primary template files
  - deployments.yaml
  - jobs.yaml
  - service-account.yaml

  # included templates must also be listed
  - configmaps_packs.yaml
  - configmaps_rbac.yaml
  - configmaps_st2-conf.yaml
  - configmaps_st2-urls.yaml
  - configmaps_st2web.yaml
  - secrets_datastore_crypto_key.yaml
  - secrets_ssh.yaml
  - secrets_st2apikeys.yaml
  - secrets_st2auth.yaml
  - secrets_st2chatops.yaml

tests:
  - ...
```

I'm not sure if we could extend helm-unittest such that pants could actually inject the included template files in that list. In any case, each test ONLY needs the files listed here (plus all `*.tpl` files and Chart.yaml). So, at least for helm-unittest, the smallest atomic unit is a template file (plus all `*.tpl` helpers files and Chart.yaml).

cognifloyd commented:
In your alternative:
```python
helm_partial_templates(name="partials")  # Uses default sources="*.tpl"
```
I don't think `*.tpl` would be correct. That is a convention, but [helm docs](https://helm.sh/docs/chart_template_guide/named_templates/#partials-and-_-files) say that helm looks for a `_` prefixed file, not the extension `*.tpl`.

Maybe default to both the prefix and the conventional suffix: `_*.tpl`

Also in your alternative:
```python
helm_kube_templates(name="kube") # Uses default sources="*.yaml"
```

That would default to `*.yaml` and `*.yml`, but would it include `NOTES.txt`? Or would that need its own target since it's not actually a "kube" template?

alonsodomin commented:
> I recently had to start writing chart unittests with helm-unittest (there are a bunch of forks. This one is presently the most maintained version: https://github.com/quintush/helm-unittest).
> 
> It would be nice to use that in the future and it would be an excellent candidate for remote execution because of how it templates things.

Great minds think alike! We also use that same plugin to run unit test and the plan is to do follow up PRs in which each of these things are added little by little. The original PR I created as a Draft included support for that Helm plugin :)

alonsodomin commented:
> In your alternative:
> 
> ```python
> helm_partial_templates(name="partials")  # Uses default sources="*.tpl"
> ```
> 
> I don't think `*.tpl` would be correct. That is a convention, but [helm docs](https://helm.sh/docs/chart_template_guide/named_templates/#partials-and-_-files) say that helm looks for a `_` prefixed file, not the extension `*.tpl`.
> 
> Maybe default to both the prefix and the conventional suffix: `_*.tpl`

You are right, Helm docs do not depend on the extension of the file. This is going to be tricky with Pants' `SourceField` as that expects a specific set of extensions, not prefixes...

> Also in your alternative:
> 
> ```python
> helm_kube_templates(name="kube") # Uses default sources="*.yaml"
> ```
> 
> That would default to `*.yaml` and `*.yml`, but would it include `NOTES.txt`? Or would that need its own target since it's not actually a "kube" template?

Yeah, the `helm_kube_templates` target name is not really a good choice. What about something like:

```python
helm_chart_templates(
  partials=["_*"],
  sources=["*.yaml", "*.yml"],
  notes="NOTES.txt"
)
```

cognifloyd commented on src/python/pants/backend/helm/target_types.py:
```diff
@@ -0,0 +1,103 @@
+    help = "If set to true, enables strict linting of this Helm chart"
+
+
+class HelmChartTarget(Target):
```
Could this be a Target Generator that generates targets:
- helm_meta_source (for `Chart.yaml`)
- helm_template_source (for each matching file in `templates/*.yaml` `templates/*.yml` and `templates/NOTES.txt`)
- helm_helpers_source (for each matching fiel in `templates/_*.tpl`)

That way, the different goals can select a subset of templates to work with and still minimize the boilerplate in BUILD. All charts have the same layout because that's what helm requires (so stronger than just a convention).

@Eric-Arellano maybe you can provide better pointers on how to acheive this (and if it's feasible)?

alonsodomin replied:
Yeah, I guess that could be a solution... I fall a bit short on how to achieve it though, as far as I know, the base class `TargetGenerator` can only generate one target type, not multiple. So any help in that front would be more than welcomed.

tdyas replied:
Raises the question from the PR description about how to model the Helm chart as targets. For example, does Pants actually need to track the template files individually?

For some language backends, the answer is yes; for example, in the JVM (Java/Scala), each source file is tracked individually so Pants can only compile the subset that actually depend on each other. Whereas the Go backend compiles by Go package, so `go_package` tracks all Go files in a package.

It seems fine to me to start with Helm support without these additional target types. If they turn out to be necessary in the future, they can be introduced then.

(As an aside, I recall that rules can also just implement `GenerateTargetsRequest` directly without using `TargetGenerator`. See `GoModTarget` in the Go backend which does not subclass `TargetGenerator`.)

cognifloyd replied:
`helm-unittest` does use individual files, not the whole chart.

alonsodomin replied:
it does at the test suite level, but a test suite can render/consume more than one template from the chart in the `templates` key of the test suite YAML file. I presume the most common expectation for any developer would be that one.

As it is now, the `HelmChartRequest` will return all the chart's sources but will not include the unit tests, so we can think of using file-level targets when adding the `helm_unittest_tests` target.

Eric-Arellano replied:
That might be good to change sooner than later. It's fairly disruptive to change something to become a target generator: it impacts the address of everything.

stuhood replied:
I do think that generators that create large networks of targets are worth being cautious about, for a lot of the same reasons as on #13488. Metadata should be able to live near the files it impacts, and if you were going to have per-template metadata, you wouldn't want it living in parent directories.

cognifloyd replied:
Hmm. I can't envision a situation that would require any per-template metadata.

tdyas commented on src/python/pants/backend/helm/target_types.py:
```diff
@@ -0,0 +1,103 @@
+
+
+# -----------------------------------------------------------------------------------------------
+# `helm_chart` target
```
Does the `Chart.lock` file need to be known to Pants at all?

alonsodomin replied:
So far we have avoided dealing with the `Chart.lock` file but some sort of implementation would be needed.

This puts forward the question of whether we want Pants to manage Helm's 3rd party dependencies or not. In Helm, the `dependencies` of a given chart are always _3rd party_, meaning that they are supposed to come from published repositories while first party dependencies (subcharts) are meant to be put inside the `charts/` folder inside the given chart.

In our experience this is somewhat messy, specially if you have a few customs charts of `library` type which are meant to be used by other, more complex, charts as it forces you to go through the `package->publish->generate-lockfiles` flow whenever you make a change to the smaller charts.

In our implementation what we do is parsing `Chart.yaml` file and infer any dependencies from it, whether they are first- or third- party, then in the rule logic that returns `HelmChart` type, we package all those subcharts under the `charts/` folder and re-render the `Chart.yaml` including the first party dependencies in it.

With that approach we avoid of publishing the first party dependencies before being able to run any `lint` or `test` goals in the other charts that reuse those.

In theory it should be possible to find a solution that satisfies both, like copying into the `charts/` folder any first party dependencies and then use `helm dependency update`/`helm dependency build` commands to let Helm resolve the third party dependencies based on the Chart.yaml.

tdyas commented on src/python/pants/backend/helm/util_rules/chart.py:
```diff
@@ -0,0 +1,253 @@
+    )
+    subset = await Get(Digest, DigestSubset, _chart_metadata_subset(source_files.snapshot.digest))
+    file_contents = await Get(DigestContents, Digest, subset)
+    return HelmChartMetadata.from_bytes(file_contents[0].content)
```
Maybe `assert len(file_contents) == 1` or raise a more descriptive exception? Could there be multiple matches on chart metadata filenames? Probably not if digest is limited to just source files of this chart, but good to assert that the `[0]` access is valid.

tdyas commented on src/python/pants/backend/helm/util_rules/chart.py:
```diff
@@ -0,0 +1,253 @@
+        f"Found {pluralize(len(first_party_subcharts), 'subchart')} as direct dependencies on Helm chart at: {field_set.address}"
+    )
+
+    # TODO Collect 3rd party chart dependencies (subcharts)
```
Track as Pants issue and put issue number here? Fine if no.

tdyas commented on src/python/pants/backend/helm/util_rules/chart.py:
```diff
@@ -0,0 +1,253 @@
+        )
+        subcharts_digest = await Get(Digest, AddPrefix(merged_subcharts, "charts"))
+
+        # TODO Update subchart dependencies in the metadata and re-render it (requires support for OCI registries and classic repositories)
```
Track as Pants issue and put issue number here? Fine if no.

tdyas commented on src/python/pants/backend/helm/util_rules/tool.py:
```diff
@@ -0,0 +1,155 @@
+class HelmProcess:
+    argv: tuple[str, ...]
+    input_digest: Digest
+    description: str = dataclasses.field(compare=False)
```
Unclear to me if uncomparable fields play nicely with the Pants engine.

cc @stuhood

stuhood replied:
Generally they should be fine, but as expected: they will be used interchangeably, which might do odd things to logging. Unless there is a specific reason to ignore fields, I'd go ahead and include them to avoid surprises.

tdyas commented on src/python/pants/backend/helm/util_rules/tool.py:
```diff
@@ -0,0 +1,155 @@
+    helm_path = os.path.join(tool_relpath, downloaded_binary.exe)
+    helm_env = _build_helm_env(cache_dir, config_dir, data_dir)
+
+    # TODO Install Global Helm plugins
+    # TODO Configure Helm classic repositories
```
Track as Pants issues and put issue numbers here? Fine if no.

alonsodomin replied:
First TODO would be give support for the `unittest` plugin, probably good that the ticket is something like _Support Helm unittest tests_ instead of just generic Helm plugin support.

The second TODO would be needed to support 3rd party dependencies.

Will create tickets for those.

stuhood reviewed (approved):
Looks good to me, thanks!

Once there is some consensus on the target layout, let's land this.

alonsodomin commented:
Addressed lastest comments from @stuhood and added additional changes to the `Chart.yaml` parser as we found that the `HelmChartRequest` rule would render a `Chart.yaml` that had missing fields compared to the originally parsed one.

Sorry for adding more code in this PR instead of using an additional PR to address that, thought it would be a good idea to include that as soon as possible.
